broadcaster/backlog/backlog.go

```diff
@@ -0,0 +1,399 @@
+			b.reset()
+			return
+		}
+	}
+
+	// tidy up lookup, count and head
+	b.removeFromLookup(start, confirmed)
+	count := b.Count() + start - confirmed - uint64(1)
+	b.messageCount.Store(count)
+	b.head.Store(newHead)
+}
+
+// removeFromLookup removes all entries from the head segment's start index to
+// the given confirmed index.
+func (b *backlog) removeFromLookup(start, end uint64) {
+	b.lookupLock.Lock()
+	defer b.lookupLock.Unlock()
+	for i := start; i <= end; i++ {
+		delete(b.lookupByIndex, i)
+	}
+}
+
+// Lookup attempts to find the backlogSegment storing the given message index.
+func (b *backlog) Lookup(i uint64) (BacklogSegment, error) {
+	b.lookupLock.RLock()
+	segment, ok := b.lookupByIndex[i]
+	b.lookupLock.RUnlock()
+	if !ok {
+		return nil, fmt.Errorf("error finding backlog segment containing message with SequenceNumber %d", i)
+	}
+
+	return segment, nil
+}
+
+// Count returns the number of messages stored within the backlog.
+func (s *backlog) Count() uint64 {
+	return s.messageCount.Load()
+}
+
+// reset removes all segments from the backlog.
+func (b *backlog) reset() {
+	b.lookupLock.Lock()
+	defer b.lookupLock.Unlock()
+	b.head = atomic.Pointer[backlogSegment]{}
+	b.tail = atomic.Pointer[backlogSegment]{}
+	b.lookupByIndex = map[uint64]*backlogSegment{}
+	b.messageCount.Store(0)
+}
+
+// BacklogSegment defines the interface for backlogSegment.
+type BacklogSegment interface {
+	Start() uint64
+	End() uint64
+	Next() BacklogSegment
+	Contains(uint64) bool
+	Messages() []*m.BroadcastFeedMessage
+}
+
+// backlogSegment stores messages up to a limit defined by the backlog. It also
+// points to the next backlogSegment in the list.
+type backlogSegment struct {
+	start           atomic.Uint64
+	end             atomic.Uint64
+	messagesLock    sync.RWMutex
+	messages        []*m.BroadcastFeedMessage
+	messageCount    atomic.Uint64
+	nextSegment     atomic.Pointer[backlogSegment]
+	previousSegment atomic.Pointer[backlogSegment]
+}
+
+// newBacklogSegment creates a backlogSegment object with an empty slice of
+// messages. It does not return an interface as it is only used inside the
+// backlog library.
+func newBacklogSegment() *backlogSegment {
+	return &backlogSegment{
+		messages: []*m.BroadcastFeedMessage{},
+	}
+}
+
+// IsBacklogSegmentNil uses the internal backlogSegment type to check if a
+// variable of type BacklogSegment is nil or not. Comparing whether an
+// interface is nil directly will not work.
+func IsBacklogSegmentNil(segment BacklogSegment) bool {
+	return segment.(*backlogSegment) == nil
+}
+
+// Start returns the first message index within the backlogSegment.
+func (s *backlogSegment) Start() uint64 {
+	return uint64(s.start.Load())
+}
+
+// End returns the last message index within the backlogSegment.
+func (s *backlogSegment) End() uint64 {
+	return uint64(s.end.Load())
+}
+
+// Next returns the next backlogSegment.
+func (s *backlogSegment) Next() BacklogSegment {
+	return s.nextSegment.Load()
```

Comment:
Maybe we could improve the nil situation a bit here, by doing:
```go
next := s.nextSegment.Load()
if next == nil {
    return nil // return a nil interface instead of a nil *backlogSegment
}
return next
```